text/0481-component-templates-co-location.md

```diff
@@ -0,0 +1,647 @@
+```
+
+First, within the `addon` folder, there is the component class and template.
+Having these files in the `addon` folder allow them to be imported from
+JavaScript (e.g. `import FooBar from 'my-addon/components/foo-bar';`) for
+subclassing. However, as mentioned above, this does not also bring along the
+component's template.
+
+To avoid developers having to also import the template, the addon component
+blueprint sets the `layout` property on the component's class. At runtime,
+when Ember fail to find a template associated with the component, it will
+then fall back to this property.
+
+Finally, it re-exports the component in the `app` folder, which is merged with
+the app's `app` folder, allowing the component to be resolved by the resolver
+and invoked from Handlebars.
+
+This workaround gets the job done, but has several major drawbacks.
+
+First, it adds a significant amount of boilerplate to addon components, which
+could be puzzling even to seasoned addon developers. In fact, the problem it
+solves is unique to addons – there is nothing wrong with using inheritance in
+apps – we just didn't think we could justify adding this amount of boilerplate
+to every app and putting them in Ember's "happy path".
+
+Second, the `layout` property in classic component uses a feature in Glimmer VM
+called "late-bound layout", in which we must assume a different template for
+each _instance_ of the component (the `layout` property can be set in `init`).
+This is fundamentally more difficult to optimize and excludes the component
+from key optimizations (and opportunities for future optimizations).
+
+Most importantly, Glimmer components do not have the equivalent of the `layout`
+property, meaning that addons (or components designed for inheritance in mind)
+cannot be written using Glimmer components today.
+
+To solve these problems once and for all, we need to change things such that
+component templates are automatically associated with their JavaScript class at
+build time, and importing the JavaScript class should result in a value that
+has this metadata associated with it, which needs to survive subclassing as
+well. Solving this problem is also essential for unblocking template imports.
+
+We believe the technical design for allowing co-location will solve this
+problem nicely.
+
+## Detailed design
+
+### High-level design
+
+We propose to allow placing a component's template adjacent to its JavaScript
+file in `app/components`. For example, for a component named `foo-bar`, it will
+be `app/components/foo-bar.js` and `app/components/foo-bar.hbs`.
+
+In addition, per the node resolution convention, we propose to allow `index`
+files inside a directory to have the equivalent semantics. In the example
+above, it could also be structured as `app/components/foo-bar/index.js` and
+`app/components/foo-bar/index.hbs`. This allows additional files related to the
+component (such as a `README.md` file) to be co-located on the filesystem.
```

Comment:
For the folder based layout, I understand why `app/components/foo-bar/index.js` represents the component’s class file; this generally makes sense from a module import perspective. I’m not sure I agree with `app/components/foo-bar/index.hbs` representing the template. This *kind of* makes semantic sense for template only components, but for non-template only components it creates two identically named files (with the exception of the extension) that both represent the component. This seems like a confusing convention to me, as JS modules typically have one entry point. Once tests are introduced to the folder structure and users update addons that provide functionality like CSS modules or Storybook tests, the index naming convention seems like it will make even less sense. If I were a new user, I'm not sure I would understand why there are two index files in this folder.

```
- foo-bar
  - index.js
  - index.hbs
  - styles.css  // index.css ??
  - tests.js
  - other.js
```

Many other component folder structures I’ve seen that utilize an `index.js` file, do so in the following way. Since Ember hooks up the component `index` for us, I don’t feel that we particularly need to adhere to an `index` naming convention, especially since it creates duplication with the template file name. That being said, I'm not totally opposed to the `index.js`/`index.hbs` proposal; I'm mostly nitpicking at this point. :)

```
- foo-bar
  - index.js  // re-exports component.js
  - component.js
  - template.hbs
```

Reply:
For a project with 100 components. This means another 100 files on disk that take up space and need to be parsed during build time even if it is a one liner. 
I actually like having the JS and HBS both be named `index.ext` 
It shows the intent that they are used *together*. :)